tests/src/python/test_qgsvectorlayertemporalproperties.py

```diff
@@ -321,32 +327,165 @@ def testExpressionMode(self):
+        self.assertTrue(layer.isValid())
+        self.assertEqual(layer.fields()[2].type(), QVariant.DateTime)
+        self.assertEqual(layer.fields()[3].type(), QVariant.DateTime)
+        # creating (point) features on a row with a one-hour duration phenomenon
+        # in a csv use:
+        #
+        #    X,Y,fid,id,t_start,t_end,text
+        #    5.02435455597371,52.6540035154413,"1","1",2020/08/26 01:00:00,2020/08/26 02:00:00,"1-2"
+        #    5.0646388842879,52.6544602897524,"2","2",2020/08/26 02:00:00,2020/08/26 03:00:00,"2-3"
+        #    5.11922979648937,52.6542319031934,"3","3",2020/08/26 03:00:00,2020/08/26 04:00:00,"3-4"
+        #    5.16930284009487,52.6546886751184,"4","4",,2020/08/26 05:00:00,"NULL-5"
+        #    5.22314077419701,52.6551454422712,"5","5",2020/08/26 05:00:00,,"5-NULL"
+        #    5.27584924115016,52.6558305840525,"6","6",2020/08/26 06:00:00,2020/08/26 07:00:00,"6-7"
+
+        # fldint and fldtxt hold the START of the temporal props (not used)
+        # fv=feature values
+        # X,Y,fldint,fldtxt,start_field,end_field,text
+        fv1 = [5.02435455597371, 52.6540035154413, 1, "1", QDateTime(QDate(2020, 8, 26), QTime(1, 0, 0)), QDateTime(QDate(2020, 8, 26), QTime(2, 0, 0)), "1-2"]
+        fv2 = [5.06463888428790, 52.6544602897524, 2, "2", QDateTime(QDate(2020, 8, 26), QTime(2, 0, 0)), QDateTime(QDate(2020, 8, 26), QTime(3, 0, 0)), "2-3"]
+        fv3 = [5.11922979648937, 52.6542319031934, 3, "3", QDateTime(QDate(2020, 8, 26), QTime(3, 0, 0)), QDateTime(QDate(2020, 8, 26), QTime(4, 0, 0)), "3-4"]
+        layer.startEditing()
+        for fv in [fv1, fv2, fv3]:
+            f = QgsFeature()
+            f.setAttributes(fv[2:])
+            f.setGeometry(QgsGeometry.fromPointXY(QgsPointXY(fv[0], fv[0])))
+            self.assertTrue(f.isValid())
+            layer.addFeature(f)
+        layer.commitChanges()
+        self.assertEqual(len(list(layer.dataProvider().getFeatures())), 3)
+
+        context = QgsVectorLayerTemporalContext()
+        context.setLayer(layer)
+        props = QgsVectorLayerTemporalProperties(enabled=True)
+        props.setMode(QgsVectorLayerTemporalProperties.ModeFeatureDateTimeStartAndEndFromFields)
+        props.setStartField('start_field')
+        props.setEndField('end_field')
+
+        # frame OUTSIDE data range, should return zero features
+        frame = QgsDateTimeRange(QDateTime(QDate(2020, 8, 25), QTime(0, 0, 0)), QDateTime(QDate(2020, 8, 25), QTime(1, 0, 0)), False, False)
+        featureRequest = QgsFeatureRequest().combineFilterExpression(props.createFilterString(context, frame))
+        result = list(layer.dataProvider().getFeatures(featureRequest))
+        self.assertEqual(len(result), 0)
+
+        frame = QgsDateTimeRange(QDateTime(QDate(2020, 8, 26), QTime(2, 0, 0)), QDateTime(QDate(2020, 8, 26), QTime(3, 0, 0)), False, False)
+        # timestep frame on exact one data range frame: 2:00-3:00
```

Comment:
what if frame range is from 1:00 - 4:00, should we expect results with 3 features or 2? maybe we need to add it as another test case at least I think it is an important one.

Reply:
IF having features with a duration of one hour, I would expect 3: those from 1-2, 2-3 and 3-4
I'll add a test case